services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreRepairTask.java

```diff
@@ -0,0 +1,255 @@
+      repairBadSystemStore(clusterName, newUnhealthySystemStoreSet, maxRepairRetry);
+    }
+  }
+
+  void repairBadSystemStore(String clusterName, Set<String> newUnhealthySystemStoreSet, int maxRepairRetry) {
+    for (int i = 0; i < maxRepairRetry; i++) {
+      Map<String, Integer> systemStoreToRepairJobVersionMap = new HashMap<>();
+      for (String systemStoreName: newUnhealthySystemStoreSet) {
+        if (!shouldContinue(clusterName)) {
+          return;
+        }
+        String pushJobId = SYSTEM_STORE_REPAIR_JOB_PREFIX + System.currentTimeMillis();
+        try {
+          Version version = UserSystemStoreLifeCycleHelper
+              .materializeSystemStore(getParentAdmin(), clusterName, systemStoreName, pushJobId);
+          systemStoreToRepairJobVersionMap.put(systemStoreName, version.getNumber());
+        } catch (Exception e) {
+          LOGGER.warn("Unable to run empty push job for store: {} in cluster: {}", systemStoreName, clusterName, e);
+        }
+      }
+      for (Map.Entry<String, Integer> entry: systemStoreToRepairJobVersionMap.entrySet()) {
+        if (pollSystemStorePushStatusUntilCompleted(clusterName, entry.getKey(), entry.getValue())) {
+          newUnhealthySystemStoreSet.remove(entry.getKey());
+          LOGGER.info("System store: {} in cluster: {} has been fixed by repair job.", entry.getKey(), clusterName);
+        }
+      }
+    }
+  }
+
+  /**
+   * This method iterates over all system stores in the given cluster by sending heartbeat and validate if heartbeat is
+   * consumed successfully by system store.
+   * At the end, it will update the count of bad meta system stores and bad DaVinci push status stores respectively.
+   */
+  void checkSystemStoresHealth(
+      String clusterName,
+      Set<String> newUnhealthySystemStoreSet,
+      Map<String, Long> systemStoreToHeartbeatTimestampMap) {
+    checkAndSendHeartbeatToSystemStores(clusterName, newUnhealthySystemStoreSet, systemStoreToHeartbeatTimestampMap);
+    try {
+      // Sleep for enough time for system store to consume heartbeat messages.
+      Thread.sleep(TimeUnit.SECONDS.toMillis(heartbeatWaitTimeSeconds));
+    } catch (InterruptedException e) {
+      LOGGER.info("Caught interrupted exception, will exit now.");
+      return;
+    }
+    checkHeartbeatFromSystemStores(clusterName, newUnhealthySystemStoreSet, systemStoreToHeartbeatTimestampMap);
+    updateBadSystemStoreCount(newUnhealthySystemStoreSet);
+  }
+
+  void checkAndSendHeartbeatToSystemStores(
+      String clusterName,
+      Set<String> newUnhealthySystemStoreSet,
+      Map<String, Long> systemStoreToHeartbeatTimestampMap) {
+    for (Store store: getParentAdmin().getAllStores(clusterName)) {
+      if (!shouldContinue(clusterName)) {
+        return;
+      }
+
+      // For user store, if corresponding system store flag is not true, it indicates system store is not created.
+      if (!VeniceSystemStoreUtils.isUserSystemStore(store.getName())) {
```

Comment:
For the regular Venice store creation, system stores will create right after, which means there is a gap in between.
Shall we check the store creation time before repairing the system stores since it is possible that the system store creation itself would take some time.